### DISCUSSION TITLE
Proxy(Blob) instead of actual Blob

### DISCUSSION THREAD

**Discussion by Ettorebigm** [pmndrs/valtio#739] [Q&A] [answered]
Hi, i was wondering if i could use Valtio as a store for Blob objects.

in one file i have this state
export const createActivityState = proxy({ activity: "", textarea: "", photo: [], photoUrl: [] });

in another i gather from < input />  user's media file, and procede as such:
createActivityState.photo.push( file );
const url = URL.createObjectURL(file)
createActivityState.photoUrl.push( url );

but when i try to access both, such as console.log , the url , as a string , "works" as it is console.logged correctly, also it is used to display the image correctly inside < img />
but the file image is substituted with Proxy(Blob) , which is not the actual file .

here attached screenshot of the console.log.

I therefore assume Valtio cannot store heavy objects as such, am i wrong / missing something  ? Or can the behaviour be achieved in other fashion ? 

Kind Regards
![Schermata 2023-06-15 alle 20 55 21](https://github.com/pmndrs/valtio/assets/134105000/13d2d09e-1905-4d8a-8c45-cc4ec867a860)

**Comment by dai-shi** [comment, marked as answer]
In fact, Valtio proxies are designed for plain objects.
In the past, we discussed about opt-in / opt-out approaches for non-plain objects. We decided for opt-out approach.
You should use `ref()` to put non-plain objects in Valtio proxies.

**Reply by Ettorebigm** [reply]
or else , to use Jotai , right?

**Reply by Ettorebigm** [reply]
what do you mean with "You should use ref() to put non-plain objects in Valtio proxies." ?

i tried to do : 
const fotoRef = createRef()
fotoRef.current = file
createActivityState.fotoRef.push(fotoRef) ,

which still console.logs Proxy(Blob)
![Schermata 2023-06-17 alle 01 37 19](https://github.com/pmndrs/valtio/assets/134105000/164b320f-0c0f-402a-abd5-6bf22e4a3790)

**Reply by dai-shi** [reply]
Oops, I mean `ref` function exported from `valtio`.

**Comment by josh231101** [comment]
The same happened to me while trying to send in a request a Blob. The error I got was 
TypeError: Failed to execute 'arrayBuffer' on 'Blob': Illegal invocation
`TypeError: Failed to execute 'arrayBuffer' on 'Blob': Illegal invocation`
So once you define your variable be sure to use **ref**. Happy coding!
